prov/efa/src/efa_errno.h

```diff
@@ -0,0 +1,119 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef EFA_ERRNO_H
+#define EFA_ERRNO_H
+
+#include <errno.h>
+
+/*
+ * These status codes correspond to EFA_IO_COMP_STATUS_xxx codes as defined by
+ * Annapurna
+ */
+enum {
+	/* Successful completion */
+	FI_EFA_OK                            = 0,
+	/* Flushed during QP destroy */
+	FI_EFA_FLUSHED                       = 1,
+	/* Internal QP error */
+	FI_EFA_LOCAL_ERROR_QP_INTERNAL_ERROR = 2,
+	/* Bad operation type */
+	FI_EFA_LOCAL_ERROR_INVALID_OP_TYPE   = 3,
+	/* Bad AH */
+	FI_EFA_LOCAL_ERROR_INVALID_AH        = 4,
+	/* LKEY not registered or does not match IOVA */
+	FI_EFA_LOCAL_ERROR_INVALID_LKEY      = 5,
+	/* Message too long */
+	FI_EFA_LOCAL_ERROR_BAD_LENGTH        = 6,
+	/* Destination ENI is down or does not run EFA */
+	FI_EFA_REMOTE_ERROR_BAD_ADDRESS      = 7,
+	/* Connection was reset by remote side */
+	FI_EFA_REMOTE_ERROR_ABORT            = 8,
+	/* Bad dest QP number (QP does not exist or is in error state) */
+	FI_EFA_REMOTE_ERROR_BAD_DEST_QPN     = 9,
+	/* Destination resource not ready (no WQEs posted on RQ) */
+	FI_EFA_REMOTE_ERROR_RNR              = 10,
+	/* Receiver SGL too short */
+	FI_EFA_REMOTE_ERROR_BAD_LENGTH       = 11,
+	/* Unexpected status returned by responder */
+	FI_EFA_REMOTE_ERROR_BAD_STATUS       = 12,
+	/* Unresponsive remote - detected locally */
+	FI_EFA_LOCAL_ERROR_UNRESP_REMOTE     = 13,
+};
+
+/*
+ * EFA-specific error codes begin here. Anything from 0 to this offset is
+ * reserved for Annapurna
+ */
+#define FI_EFA_ERRNO_OFFSET 0x1000
+
+/* Be careful changing the order of identifiers in this enum */
+enum {
+	FI_EFA_ERR_OTHER = FI_EFA_ERRNO_OFFSET, /* Unknown/other error */
+	FI_EFA_ERR_DEPRECATED_PKT_TYPE,
+	FI_EFA_ERR_INVALID_PKT_TYPE,
+	FI_EFA_ERR_UNKNOWN_PKT_TYPE,
+	FI_EFA_ERR_PKT_POST,
+	FI_EFA_ERR_PKT_POST_OR_QUEUE,
+	FI_EFA_ERR_PKT_SEND_QUEUED,
+	FI_EFA_ERR_PKT_PROC_MSGRTM,
+	FI_EFA_ERR_PKT_PROC_TAGRTM,
+	FI_EFA_ERR_PKT_ALREADY_PROCESSED,
+	FI_EFA_ERR_OOM,
+	FI_EFA_ERR_MR_CLOSE,
+	FI_EFA_ERR_MR_DEREG,
+	FI_EFA_ERR_RX_ENTRY_COPY,
+	FI_EFA_ERR_RX_ENTRIES_EXHAUSTED,
+	FI_EFA_ERR_TX_ENTRIES_EXHAUSTED,
+	FI_EFA_ERR_AV_INSERT,
+	FI_EFA_ERR_RMA_ADDR,
+	FI_EFA_ERR_BULK_POST_RECV_BUF,
+	FI_EFA_ERR_PEER_HANDSHAKE,
+	FI_EFA_ERR_POST_FLUSH,
+	FI_EFA_ERR_RTM_MISMATCH,
+	FI_EFA_ERR_READ_POST,
+	FI_EFA_ERR_RDMA_READ_POST,
+	FI_EFA_ERR_RDMA_READ_POST_OR_QUEUE,
+	FI_EFA_ERR_INVALID_DATATYPE,
+	FI_EFA_ERR_WRITE_COMP,
+	FI_EFA_ERR_SHM_INTERNAL_ERROR,
+	FI_EFA_ERR_SHM_CQ_ENTRY,
+	FI_EFA_ERRNO_MAX                        /* Dummy errno for upper bound */
+};
```

Comment:
Feedback appreciated here. I want to make sure these are as comprehensive and well-defined as possible before publishing.